{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[![Jupyter Notebook](https://img.shields.io/badge/Source%20on%20GitHub-orange)](https://github.com/laminlabs/lamin-usecases/blob/main/docs/spatial2.ipynb)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Spatial RNA-seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, you'll learn how to manage spatial datasets:\n",
    "\n",
    "1. query & analyze spatial datasets (![spatial1/4](https://img.shields.io/badge/spatial1/4-lightgrey))\n",
    "2. create and share interactive visualizations with vitessce ([![spatial2/4](https://img.shields.io/badge/spatial2/4-lightgrey)](/spatial2))\n",
    "3. curate and ingest spatial data (![spatial3/4](https://img.shields.io/badge/spatial3/4-lightgrey))\n",
    "4. load the collection into memory & train a ML model ([![spatial3/4](https://img.shields.io/badge/spatial4/4-lightgrey)](/spatial3))\n",
    "\n",
    "\n",
    "\n",
    "```{toctree}\n",
    ":maxdepth: 1\n",
    ":hidden:\n",
    "\n",
    "spatial2\n",
    "spatial3\n",
    "spatial4\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Spatial omics data integrates molecular profiling (e.g., transcriptomics, proteomics) with spatial information, preserving the spatial organization of cells and tissues.\n",
    "It enables high-resolution mapping of molecular activity within biological contexts, crucial for understanding cellular interactions and microenvironments.\n",
    "\n",
    "Many different spatial technologies such as multiplexed imaging, spatial transcriptomics, spatial proteomics, whole slide imaging, spatial metabolomics, and 3D tissue reconstruction exist which can all be stored in the [SpatialData](https://github.com/scverse/spatialdata) data framework.\n",
    "For more details we refer to the original publication:\n",
    "\n",
    "Marconato, L., Palla, G., Yamauchi, K.A. et al. SpatialData: an open and universal data framework for spatial omics. Nat Methods 22, 58â€“62 (2025). [https://doi.org/10.1038/s41592-024-02212-x](https://doi.org/10.1038/s41592-024-02212-x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```{note}\n",
    "A collection of curated spatial datasets in SpatialData format is available on the [scverse/spatialdata-db instance](https://lamin.ai/scverse/spatialdata-db).\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```{dropdown} spatial data vs SpatialData terminology\n",
    "When we mention spatial data, we refer to data from spatial assays, such as spatial transcriptomics or proteomics, that includes spatial coordinates to represent the organization of molecular features in tissue.\n",
    "When we refer SpatialData, we mean spatial omics data stored in the scverse SpatialData framework.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Query and analyze spatial data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "# pip install lamindb spatialdata spatialdata-plot squidpy\n",
    "!lamin init --storage ./test-spatial --modules bionty"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import lamindb as ln\n",
    "import scanpy as sc\n",
    "import spatialdata as sd\n",
    "import spatialdata_plot as pl\n",
    "import squidpy as sq\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "ln.track()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query by biological metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll work with a human lung cancer dataset generated using 10x Genomics Xenium platform and available in a public instance. \n",
    "This FFPE (formalin-fixed paraffin-embedded) tissue sample includes spatial gene expression profiles.\n",
    "\n",
    "Create the central query object of our public [lamindata instance](https://lamin.ai/laminlabs/lamindata):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "db = ln.DB(\"laminlabs/lamindata\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now query the database to find Xenium datasets from lung tissue:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "all_xenium_data = db.Artifact.filter(\n",
    "    assay=\"Xenium Spatial Gene Expression\", tissue=\"lung\"\n",
    ")\n",
    "\n",
    "all_xenium_data.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyze spatial data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "sdata = all_xenium_data[0].load()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Spatial data datasets stored as SpatialData objects can easily be examined and analyzed through the SpatialData framework, [squidpy](https://github.com/scverse/squidpy), and [scanpy](https://github.com/scverse/scanpy).\n",
    "\n",
    "We use [spatialdata-plot](https://github.com/scverse/spatialdata-plot) to get an overview of the dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "axes = plt.subplots(1, 2, figsize=(10, 10))[1].flatten()\n",
    "sdata.pl.render_images(\"he_image\", scale=\"scale4\").pl.show(\n",
    "    ax=axes[0], title=\"H&E image\"\n",
    ")\n",
    "sdata.pl.render_images(\"morphology_focus\", scale=\"scale4\").pl.show(\n",
    "    ax=axes[1], title=\"Morphology image\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can visualize the segmentations masks by rendering the shapes from the SpatialData object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def crop0(x):\n",
    "    return sd.bounding_box_query(\n",
    "        x,\n",
    "        min_coordinate=[20000, 7000],\n",
    "        max_coordinate=[22000, 7500],\n",
    "        axes=(\"x\", \"y\"),\n",
    "        target_coordinate_system=\"global\",\n",
    "    )\n",
    "\n",
    "\n",
    "crop0(sdata).pl.render_images(\"he_image\", scale=\"scale2\").pl.render_shapes(\n",
    "    \"cell_boundaries\", fill_alpha=0, outline_alpha=0.5\n",
    ").pl.show(\n",
    "    figsize=(15, 10), title=\"H&E image & cell boundaries\", coordinate_systems=\"global\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For any Xenium analysis we would use the `AnnData` object, which contains the count matrix, cell and gene annotations.\n",
    "It is stored in the `spatialdata.tables` slot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "adata = sdata.tables[\"table\"]\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quality control metrics can be calculated using `scanpy.pp.calculate_qc_metrics`. \n",
    "Cells and genes can be filtered based on minimum thresholds using `scanpy.pp.filter_cells` and `scanpy.pp.filter_genes`.\n",
    "\n",
    "For normalization and dimensionality reduction, standard [scanpy](https://github.com/scverse/scanpy) workflows include:\n",
    "- `scanpy.pp.normalize_total` for library size normalization\n",
    "- `scanpy.pp.log1p` for log transformation\n",
    "- `scanpy.pp.pca` for principal component analysis\n",
    "- `scanpy.pp.neighbors` for neighborhood graph computation\n",
    "\n",
    "For this analysis, we use precomputed Leiden clustering results rather than running the full preprocessing pipeline. \n",
    "For a full tutorial on how to perform analysis of Xenium data, we refer to [squidpy's Xenium tutorial](https://squidpy.readthedocs.io/en/stable/notebooks/tutorials/tutorial_xenium.html)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualize annotation on UMAP and spatial coordinates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\n",
    "        \"total_counts\",\n",
    "        \"n_genes_by_counts\",\n",
    "        \"leiden\",\n",
    "    ],\n",
    "    wspace=0.1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=[\n",
    "        \"leiden\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "ln.finish()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
