{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[![Jupyter Notebook](https://img.shields.io/badge/Source%20on%20GitHub-orange)](https://github.com/laminlabs/lamin-usecases/blob/main/docs/spatial2.ipynb)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Spatial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, you'll learn how to manage spatial datasets:\n",
    "\n",
    "1. query & analyze spatial datasets (![spatial1/4](https://img.shields.io/badge/spatial1/4-lightgrey))\n",
    "2. create and share interactive visualizations with vitessce ([![spatial2/4](https://img.shields.io/badge/spatial2/4-lightgrey)](/spatial2))\n",
    "3. curate and ingest spatial data (![spatial3/4](https://img.shields.io/badge/spatial3/4-lightgrey))\n",
    "4. load the collection into memory & train a ML model ([![spatial3/4](https://img.shields.io/badge/spatial4/4-lightgrey)](/spatial3))\n",
    "\n",
    "\n",
    "\n",
    "```{toctree}\n",
    ":maxdepth: 1\n",
    ":hidden:\n",
    "\n",
    "spatial2\n",
    "spatial3\n",
    "spatial4\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Spatial omics data integrates molecular profiling (e.g., transcriptomics, proteomics) with spatial information, preserving the spatial organization of cells and tissues.\n",
    "It enables high-resolution mapping of molecular activity within biological contexts, crucial for understanding cellular interactions and microenvironments.\n",
    "\n",
    "Many different spatial technologies such as multiplexed imaging, spatial transcriptomics, spatial proteomics, whole slide imaging, spatial metabolomics, and 3D tissue reconstruction exist which can all be stored in the [SpatialData](https://github.com/scverse/spatialdata) data framework.\n",
    "For more details we refer to the original publication:\n",
    "\n",
    "Marconato, L., Palla, G., Yamauchi, K.A. et al. SpatialData: an open and universal data framework for spatial omics. Nat Methods 22, 58â€“62 (2025). [https://doi.org/10.1038/s41592-024-02212-x](https://doi.org/10.1038/s41592-024-02212-x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```{note}\n",
    "A collection of curated spatial datasets in SpatialData format is available on the [scverse/spatialdata-db instance](https://lamin.ai/scverse/spatialdata-db).\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```{dropdown} spatial data vs SpatialData terminology\n",
    "When we mention spatial data, we refer to data from spatial assays, such as spatial transcriptomics or proteomics, that includes spatial coordinates to represent the organization of molecular features in tissue.\n",
    "When we refer SpatialData, we mean spatial omics data stored in the scverse SpatialData framework.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Query and analyze spatial data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pip install 'lamindb[jupyter,bionty]' spatialdata spatialdata-plot\n",
    "!lamin init --storage ./test-spatial --modules bionty"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import lamindb as ln\n",
    "import scanpy as sc\n",
    "import spatialdata as sd\n",
    "import spatialdata_plot as pl\n",
    "import squidpy as sq\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "ln.track()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query by biological metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll work with a human lung cancer dataset generated using 10x Genomics Xenium platform and available in a public instance. \n",
    "This FFPE (formalin-fixed paraffin-embedded) tissue sample includes spatial gene expression profiles.\n",
    "\n",
    "Let's query the database to find Xenium datasets from lung tissue:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "all_xenium_data = (\n",
    "    ln.Artifact.connect(\"laminlabs/lamindata\")\n",
    "    .filter(assay=\"Xenium Spatial Gene Expression\", tissue=\"lung\")\n",
    ")\n",
    "all_xenium_data.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyze spatial data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Spatial data datasets stored as SpatialData objects can easily be examined and analyzed through the SpatialData framework, [squidpy](https://github.com/scverse/squidpy), and [scanpy](https://github.com/scverse/scanpy):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "sdata = all_xenium_data[0].load()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use `spatialdata-plot` to get an overview of the dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "axes = plt.subplots(1, 2, figsize=(10, 10))[1].flatten()\n",
    "sdata.pl.render_images(\"he_image\", scale=\"scale4\").pl.show(ax=axes[0], title=\"H&E image\")\n",
    "sdata.pl.render_images(\"morphology_focus\", scale=\"scale4\").pl.show(ax=axes[1], title=\"Morphology image\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can visualize the segmentations masks by rendering the shapes from the SpatialData object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def crop0(x):\n",
    "    return sd.bounding_box_query(\n",
    "        x,\n",
    "        min_coordinate=[20000, 7000],\n",
    "        max_coordinate=[22000, 7500],\n",
    "        axes=(\"x\", \"y\"),\n",
    "        target_coordinate_system=\"global\",\n",
    "    )\n",
    "\n",
    "crop0(sdata).pl.render_images(\"he_image\", scale=\"scale2\").pl.render_shapes(\n",
    "    \"cell_boundaries\", \n",
    "    fill_alpha=0, \n",
    "    outline_alpha=0.5\n",
    "    ).pl.show(figsize=(15, 10),\n",
    "              title=\"H&E image & cell boundaries\", \n",
    "              coordinate_systems=\"global\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For any Xenium analysis we would use the `AnnData` object, which contains the count matrix, cell and gene annotations.\n",
    "It is stored in the `spatialdata.tables` slot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "adata = sdata.tables['table']\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the quality control metrics on the AnnData object using `scanpy.pp.calculate_qc_metrics`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(adata, percent_top=(10, 20, 50, 150), inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The percentage of control probes and control codewords can be calculated from the `obs` slot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "cprobes = (\n",
    "    adata.obs[\"control_probe_counts\"].sum() / adata.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "cwords = (\n",
    "    adata.obs[\"control_codeword_counts\"].sum() / adata.obs[\"total_counts\"].sum() * 100\n",
    ")\n",
    "print(f\"Negative DNA probe count %: {cprobes}\")\n",
    "print(f\"Negative decoding count %: {cwords}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the distribution of total transcripts per cell, unique transcripts per cell, area of segmented cells and the ratio of nuclei area to their cells:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 4, figsize=(15, 4))\n",
    "\n",
    "axs[0].set_title(\"Total transcripts per cell\")\n",
    "axs[0].hist(adata.obs[\"total_counts\"], bins=100)\n",
    "\n",
    "axs[1].set_title(\"Unique transcripts per cell\")\n",
    "axs[1].hist(adata.obs[\"n_genes_by_counts\"], bins=100)\n",
    "\n",
    "axs[2].set_title(\"Area of segmented cells\")\n",
    "axs[2].hist(adata.obs[\"cell_area\"], bins=100)\n",
    "\n",
    "axs[3].set_title(\"Nucleus ratio\")\n",
    "axs[3].hist(adata.obs[\"nucleus_area\"] / adata.obs[\"cell_area\"], bins=100)\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter the cells based on the minimum number of counts required using `scanpy.pp.filter_cells` and the genes based on the minimum number of cells required with `scanpy.pp.filter_genes`. \n",
    "The parameters for the both were specified based on the plots above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(adata, min_counts=10)\n",
    "sc.pp.filter_genes(adata, min_cells=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Normalize the total counts per cell using `scanpy.pp.normalize_total`, apply log transformation with `scanpy.pp.log1p`, perform principal component analysis using `scanpy.pp.pca`, and compute a neighborhood graph of observations with `scanpy.pp.neighbors`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers[\"counts\"] = adata.X.copy()\n",
    "sc.pp.normalize_total(adata, inplace=True)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.pca(adata)\n",
    "sc.pp.neighbors(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compute a UMAP embedding of the neighborhood graph using `scanpy.tl.umap` and cluster cells into groups using `scanpy.tl.leiden`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.umap(adata)\n",
    "sc.tl.leiden(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualize annotation on UMAP and spatial coordinates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color=[\n",
    "        \"total_counts\",\n",
    "        \"n_genes_by_counts\",\n",
    "        \"leiden\",\n",
    "    ],\n",
    "    wspace=0.1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    adata,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=[\n",
    "        \"leiden\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For a full tutorial on how to perform analysis of Xenium data, we refer to [squidpy's Xenium tutorial](https://squidpy.readthedocs.io/en/stable/notebooks/tutorials/tutorial_xenium.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "ln.finish()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
